### ISSUE TITLE
Elide unchanged lines from string diffs

### ISSUE THREAD

**Issue by evanrelf** [Gabriel439/nix-diff#42]
When a long string differs between derivations - usually some long shell script written in a multi-line string - `nix-diff` prints the entire string contents, even if only a few lines of that long string have changed.

It would be nice if the parts in common were elided, like `diff` does with files.

---

I've reproduced this in a gist to illustrate: https://gist.github.com/evanrelf/2b1b1772c5f83ca102c89d29e4839aa3

I have two derivations that write similar (long) shell scripts using `writeShellScript`. Only one line is different between the Nix source code and the resulting shell scripts.

`diff -u a.nix b.nix` and `diff -u $(nix-build a.nix) $(nix-build b.nix)` are both very short; they elide parts in common.

`nix-diff $(nix-instantiate a.nix) $(nix-instantiate b.nix)` prints the whole string.

**Comment by SuperSandro2000** [comment]
I am really interested in this feature and it is really the only bummer in nix-diff right now.

Maybe the diff algorithm should be replaced by something smarter in general like https://hackage.haskell.org/package/Diff ?

**Comment by Gabriella439** [comment]
I'd accept a pull request to fix this, but I likely won't have the time to implement this myself

**Comment by SuperSandro2000** [comment]
I tried getting that implemented but it turns out my none existing Haskell knowledge made it impossible.

**Comment by evanrelf** [comment]
@SuperSandro2000 I don't believe the `Diff` library is any more intelligent than the `patience` library `nix-diff` uses today. In fact, [I replaced `Diff` with `patience` a while ago](https://github.com/Gabriella439/nix-diff/pull/33) because it _improved_ the diffs! 🙂

**Comment by SuperSandro2000** [comment]
Good to know that!
